### ISSUE TITLE
Very slow recording of TFRecordWriter with tf.data.Dataset.shard()

### ISSUE THREAD

**Issue by theonekeyg** [tensorflow/tensorflow#39197]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.2.0-dev20200412
- Python version: 3.6.10

**Describe the current behavior**
I'm trying to convert my dataset to multiple *.tfrecord files. My initial idea was to compile a `tf.data.Dataset` with all data of a certain class, and then split it into multiple datasets using `tf.data.Dataset.shard(...)`, which produced terrible recording to disk speed. 

As a workaround i split and compiled the data first into multiple `tf.data.Dataset` and recorded them as it is, which produced much better results.

Code to reproduce: https://gist.github.com/theonekeyg/68b3df515ce872d70611f92a857895a1
The exact data that i used can be obtained from this Kaggle competition: https://www.kaggle.com/c/alaska2-image-steganalysis/data

To be more specific, with `tf.data.Dataset.shard(...)` approach, to process 3000 samples it took ~35 minutes, and ~1.2 minutes with second approach.

Is that an expected behavior, that `ShardDataset` takes so long to compile/write to .tfrecord?

**Comment by amahendrakar** [comment]
@theonekeyg,
Due to the size of the dataset, I am unable to reproduce the issue. Could you please provide a sample dataset of a smaller size, so that we can look into this. Thanks!

**Comment by theonekeyg** [comment]
Hi, sure. One thing i found in particular while i was making a slice of a dataset, is that the performance issue isn't that clear or completely doesn't appear on small datasets. I tried to keep the slice as small as i could but so it still could clearly produce the performance issue, i also put the scripts to reproduce the issue to .zip archive: https://drive.google.com/file/d/1PkZjEJzUn3ioXIoNWgXDtSU9zMmsQAkk/view?usp=sharing

Hopefully you can see as i see, and also, according to my experience, the larger the dataset, the clearer the issue, so if the issue won't appear, i recommend trying it on bigger dataset, or i could upload larger sample of ALASKA2 dataset.

**Comment by leoffx** [comment]
I'm having the same issue, and noticed what you told is true. If I'm writing a TFrecord with ~16000 images in it, I get 2 it/s and writing 4040 images gives me 6 it/s.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.2. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/85f404dbae4ec1679cf756e8199b94bb/39197.ipynb). Thanks!

**Comment by jsimsa** [comment]
Why not use the method suggested by the [official documentation](https://www.tensorflow.org/api_docs/python/tf/data/experimental/TFRecordWriter)?

**Comment by theonekeyg** [comment]
It's not about what approach to use, although i tested the method from official documentation, it produces about the same speed as my second approach. I am curious if that's a known issue, that ShardDataset produces really bad speed at writing to .tfrecord?

**Comment by jsimsa** [comment]
The [official documentation](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#shard):

"Generally it is best if the shard operator is used early in the dataset pipeline. For example, when reading from a set of TFRecord files, shard before converting the dataset to input samples. ..."

This is because `shard` will evaluate the entire upstream input pipeline filtering out `(num_shards - 1) / num_shards` of the data. In your example, you are running the input pipeline inside of a for loop, which means that you will preprocess the entire input pipeline `num_shards` many times.

You could use `window` instead of `shard` to divide up the dataset into chunks and save each chunk separately along the following lines:

```
dataset = ...
dataset = dataset.window(ELEMENTS_PER_FILE)
dataset = dataset.enumerate()

def write_data(i, dataset):
  out_path = os.path.join(output_folder, f'{folder}{i+1}.tfrecord')
  dump_dataset(dataset, out_path)
  return out_path

dataset = dataset.map(write_dataset, num_parallel_calls=tf.data.experimental.AUTOTUNE)
```

If you get rid of using `from_generator` in your `dump_dataset` method, you would be able to do the writing in parallel as well (which is additional inefficiency of your program -- writing a unnecessarily serialized on Python GIL).

**Comment by theonekeyg** [comment]
It's clear to me now, thanks for the thoughtful explanation :) @jsimsa

**Comment by mpicci** [comment]
Hello,
@jsimsa do you have a working example?

Thank you

**Comment by st3inum** [comment]
In my case I follow the following [blog](https://towardsdatascience.com/how-to-build-efficient-audio-data-pipelines-with-tensorflow-2-0-b3133474c3c1) and use [this code](https://gist.githubusercontent.com/dschwertfeger/0c787c0af781c28720c28f1bebc9a126/raw/37dc7a5be0cc79ecb9809898a8e460133b936f76/convert.py).

I have 300000 audio files(each being a 10s 48KHz wav file) approximately. It will take approximately 30hour to convert all the wav files to tfrecord. Am I doing something wrong? or it is normal? 

Note: I've seen the disk usage and CPU usage. It shows that it is using only one core.

**Comment by parismic** [comment]
> **UPD:** **I solved the issue with `multiprocessing`, with it I reduced my convertion time 32x less. The only thing I did is, divide the task among the cores.**

Could you expand on this statement. How did you do it? How did you divide the data?
